### PR TITLE
Poi3dLayer improvements

### PR DESCRIPTION
**Pull request by Gustl22**
- add ZoomLimiter
- Cover areas with models
- Allow render theme ~in areas~ _EDIT: for all elements_ (return false)

(Requires #609)

### CODE REVIEW

**Comment by Gustl22** [comment]
Ok there sometimes appear this error:
```
Exception in thread "LWJGL Application" java.lang.ArrayIndexOutOfBoundsException: 0
	at org.oscim.gdx.poi3d.Poi3DLayer.onMapEvent(Poi3DLayer.java:420)
	at org.oscim.map.Map$2.tell(Map.java:153)
	at org.oscim.map.Map$2.tell(Map.java:149)
	at org.oscim.event.EventDispatcher.fire(EventDispatcher.java:66)
	at org.oscim.map.Map.prepareFrame(Map.java:414)
	at org.oscim.gdx.GdxMap$MapAdapter.access$000(GdxMap.java:146)
	at org.oscim.gdx.GdxMap$MapAdapter$1.run(GdxMap.java:171)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication.executeRunnables(LwjglApplication.java:259)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication.mainLoop(LwjglApplication.java:210)
	at com.badlogic.gdx.backends.lwjgl.LwjglApplication$1.run(LwjglApplication.java:126)
```
Needs fixing before merge, sry...

**Comment by Gustl22** [comment]
Now fixed issues:
- handle empty ancestor tiles,
- option for hiding elements rendered with theme, 
- render models on line evenly spread

File: vtm-gdx-poi3d/src/org/oscim/gdx/poi3d/Poi3DLayer.java [hunk 1]
```diff
@@ -112,14 +133,95 @@ public boolean process(MapTile tile, RenderBuckets buckets, MapElement element)
...
+                                        polyPoints.add(y);
+                                    }
+                                    tmpY += modelDistPix;
+                                }
+                                box.xmin += modelDistPix;
+                            }
+                        } else {
+                            // Place models on a line with a gap of MODEL_DISTANCE
+                            float[] p1 = new float[2]; // TODO avoid initialization
+                            float[] p2 = new float[2];
```
Comment (devemux86) [inline vtm-gdx-poi3d/src/org/oscim/gdx/poi3d/Poi3DLayer.java]: Indeed those should move in class level to avoid continuous re-creation with each call.

Comment (Gustl22) [reply vtm-gdx-poi3d/src/org/oscim/gdx/poi3d/Poi3DLayer.java]: I made some tests and I cannot clearly confirm that object creation has influence on performance, in this case. Furthermore `process()` is not synchronized and can keep up with the alternative implementation (which needs to be synchronized).

**Comment by devemux86** [comment]
Can you post a new PR rebased on latest master?

In future we should try to avoid interdependent PRs. In case of dependencies should wait the first ones to be reviewed / merged first.

They complicate the review / merge process, since first PRs may need changes and those cannot be applied easily to last ones, leading to rebase / merge failure.

**Comment by Gustl22** [comment]
> Can you post a new PR rebased on latest master?

Sure!

> They complicate the review / merge process, since first PRs may need changes and those cannot be applied easily to last ones, leading to rebase / merge failure.

Yes, I agree. For dependent PRs, there is no good workaround in git.

**Comment by devemux86** [comment]
> Global variables or local instantiation

Actually there isn't a global rule, as with most cases.
Depending the visibility of the variables and their creation frequency, usually need to examine carefully the performance of both time and memory per case.

**Comment by Gustl22** [comment]
Ok thanks for the info, I haven't considered the memory.

**Pull request closed without merging**